sopel/modules/clock.py

```diff
@@ -55,72 +57,89 @@ def setup(bot):
     bot.config.define_section('clock', TimeSection)
 
 
-@commands('t', 'time')
-@example('.t America/New_York')
+@module.commands('t', 'time')
+@module.example('.t America/New_York')
+@module.example('.t Exirel')
+@module.example('.t #sopel')
 def f_time(bot, trigger):
-    """Returns the current time."""
+    """Returns the current time.
+
+    The command takes an optional parameter: it can be either a timezone, a
+    nick, or a channel.
+    """
     if trigger.group(2):
-        zone = get_timezone(bot.db, bot.config, trigger.group(2).strip(), None, None)
-        if not zone:
-            bot.say('Could not find timezone %s.' % trigger.group(2).strip())
-            return
+        argument = trigger.group(2).strip()
+        channel_or_nick = tools.Identifier(argument)
+
+        if channel_or_nick in bot.users:
+            zone = get_nick_timezone(bot.db, channel_or_nick)
+            if zone is None:
+                bot.say('Could not find timezone for nick %s.' % argument)
+                return
+        elif channel_or_nick in bot.channels:
+            zone = get_channel_timezone(bot.db, channel_or_nick)
+            if zone is None:
+                bot.say('Could not find timezone for channel %s.' % argument)
+                return
+        else:
+            try:
+                zone = validate_timezone(argument)
+            except ValueError:
+                bot.say('Could not find timezone %s.' % argument)
+                return
     else:
         zone = get_timezone(bot.db, bot.config, None, trigger.nick,
                             trigger.sender)
+
     time = format_time(bot.db, bot.config, zone, trigger.nick, trigger.sender)
     bot.say(time)
 
 
-@commands('settz', 'settimezone')
-@example('.settz America/New_York')
+@module.commands('settz', 'settimezone')
+@module.example('.settz America/New_York')
 def update_user(bot, trigger):
     """
     Set your preferred time zone. Most timezones will work, but it's best to
     use one from <https://sopel.chat/tz>.
     """
-    if not pytz:
-        bot.reply("Sorry, I don't have timezone support installed.")
+    tz = trigger.group(2)
+    if not tz:
+        bot.reply("What timezone do you want to set? Try one from "
+                  "https://sopel.chat/tz")
+        return
+
+    if tz not in pytz.all_timezones:
+        bot.reply("I don't know that time zone. Try one from "
+                  "https://sopel.chat/tz")
+        return
+
+    bot.db.set_nick_value(trigger.nick, 'timezone', tz)
+    if len(tz) < 7:
```

Comment:
Shouldn't this be `<= 4`?

Though even that would still yield false warnings for e.g. `Iran`. The current value emits the warning for perfectly valid zone names like `Israel`, `Egypt`, and others, which is not great. (As with the above validity check, this too could be done in a subsequent PR, but we might as well do it now while we are touching the code anyway.)